pkgs/top-level/stage.nix

```diff
@@ -176,155 +176,104 @@ let
       ((config.packageOverrides or (super: {})) super);
 
   # Convenience attributes for instantitating package sets. Each of
-  # these will instantiate a new version of allPackages. Currently the
-  # following package sets are provided:
-  #
-  # - pkgsCross.<system> where system is a member of lib.systems.examples
-  # - pkgsMusl
-  # - pkgsi686Linux
-  otherPackageSets = self: super: {
+  # these will instantiate a new version of allPackages.
+  otherPackageSets = let
+    mkPkgs = name: fn: nixpkgsFun (prevArgs: let nixpkgsArgs = fn prevArgs; in nixpkgsArgs // {
+      overlays = [
+        (self': super': {
+          "${name}" = super';
+        })
+      ] ++ nixpkgsArgs.overlays or [] ++ prevArgs.overlays or [];
+    });
+    # This is always cross.
+    mkCrossPkgs = name: crossAttrs: mkPkgs name (prevArgs: {
+      crossSystem =
+        (lib.systems.systemToAttrs (lib.defaultTo prevArgs.localSystem prevArgs.crossSystem or null)) // crossAttrs;
+    });
+    # This is only cross when we are already cross, otherwise local.
+    # For the case of "native cross", i.e. pkgsCross.gnu64 on a x86_64-linux system, we need to adjust **both**
+    # localSystem **and** crossSystem, otherwise they're out of sync.
+    mkHybridPkgs = name: hybridAttrs: mkPkgs name (prevArgs: let
+        newSystem = (lib.systems.systemToAttrs (lib.defaultTo prevArgs.localSystem prevArgs.crossSystem or null)) // hybridAttrs;
+      in lib.optionalAttrs (stdenv.hostPlatform == stdenv.buildPlatform) { localSystem = newSystem; }
+      // lib.optionalAttrs (prevArgs ? crossSystem) { crossSystem = newSystem; }
```

Comment:
Cross-ness is not an intrinsic, but derived property of a build+host configuration.
By differentiating on the presence of `crossSystem`, we're extending the configuration space from build+host to build+host+cross, which complicates the interface and/or user interface of Nixpkgs as a function.

Concretely, a configuration where the `localSystem` and `crossSystem` are identical values should be indistinguishable from a configuration where only `localSystem` was passed.
I believe this is currently not entirely the case, which is a bug.

This particularly line with expression `prevArgs?crossSystem` seems either affected by or prone to that bug.